src/System.Threading.Thread/tests/ThreadTests.cs

```diff
@@ -427,6 +429,56 @@ public static void CurrentPrincipalTest()
             });
         }
 
+        [Fact]
+        public static void CurrentPrincipalContextFlowTest()
+        {
+            ThreadTestHelpers.RunTestInBackgroundThread(async () =>
+            {
+                Thread.CurrentPrincipal = new ClaimsPrincipal();
+
+                await Task.Run(async() => {
+
+                    Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal);
+
+                    await Task.Run(async() => 
+                    {
+                        Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal);
+
+                        Thread.CurrentPrincipal = new GenericPrincipal(new GenericIdentity("name"), new string[0]);
+
+                        await Task.Run(() =>
+                        {
+                            Assert.IsType<GenericPrincipal>(Thread.CurrentPrincipal);
+                        });
+
+                        Assert.IsType<GenericPrincipal>(Thread.CurrentPrincipal);
+                    });
+
+                    Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal);
+                });
+
+                Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal);
+            });
+        }
+
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "ExecutionContext.SuppressFlow() doesn't work for Task.Factory.StartNew")]
+        public static void CurrentPrincipalContextFlowTest_NotFlow()
+        {
+            ThreadTestHelpers.RunTestInBackgroundThread(async () =>
+            {
+                Thread.CurrentPrincipal = new ClaimsPrincipal();
+
+                // TaskCreationOptions.LongRunning to have fresh new thread with clean execution context
+                await Task.Factory.StartNew(() => Assert.IsType<ClaimsPrincipal>(Thread.CurrentPrincipal), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+
+                ExecutionContext.SuppressFlow();
+
+                // TaskCreationOptions.LongRunning to have fresh new thread with clean execution context
+                await Task.Factory.StartNew(() => Assert.True(Thread.CurrentPrincipal is null), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+            });
```

Comment:
Should we test that re-enabling flow works? From the docs seems like that's `ExecutionContext.SuppressFlow().Undo()` ?

Reply:
You're right, however we cannot undo because we get exception, undo must be called on thread called suppress(continuation is on other thread pool thread), so I preferred avoid risks and run test in custom process, is it a good idea?

Reply:
> however we cannot undo because we get exception

You can, you just can't around an await, so instead of:
```C#
using (ExecutionContext.SuppressFlow())
{
    await Task.*;
}
```
you'd do:
```C#
Task t;
using (ExecutionContext.SuppressFlow())
{
    t = Task.*;
}
await t;
```

Reply:
Good point!